Fixes #26753 - react-router foundation #3

glekner opened the pull request:
add `onClick` prop to Bookmarks to disable`Turbolinks.visit` when not needed.


<!--
Simple description of what is fixed/introduced.

Prerequisites for testing:
* VMware cluster
* Two smart proxies

Tests needed (anticipated impacts):
* Hosts list - mainly power column
* Power status on host page
-->

<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [Translating section in the guide]
* Suggest prerequisites for testing and testing scenarios following example above.
(https://projects.theforeman.org/projects/foreman/wiki/Translating)
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* Be nice and respectful

We are running bots that will poke you if you miss an item from the list :-)

--->

glekner committed:
Fixes #26753 - react-router foundation #3 - Bookmarks

sharvit commented on webpack/assets/javascripts/react_app/components/SearchBar/SearchBarHelpers.js:
```diff
@@ -1,8 +1,7 @@
-import URI from 'urijs';
+import { getURI, changeQuery } from '../Pagination/PaginationHelper';
```
* Isn't it odd you are importing `getURI` and `changeQuery` from `PaginationHelper`?
* isn't it suppose to use `react-router` now for those type of operations?

glekner replied:
this function is for the current views not for the `react-router` ones. either way I think its best to move all URI functions to a dedicated helper.

glekner commented:
- Moved all URI helpers to dedicated file.

sharvit commented:
> * Moved all URI helpers to dedicated file.

@glekner notice @MariaAga commit that creates the same helper with code moved from `katello`:
https://github.com/theforeman/foreman/commit/6730cf2a5b440021373a5687dcc2adacfd61bd1d#diff-5583b0c21a9139208c8282c8618869a2

glekner commented:
Summarizing changes in this PR:

- moved all URI funcs/helpers to `urlHelpers`
- added Bookmarks functionality for `react-router`
- added missing css classes for collapse nav state
- added empty react controller view and added it to permissions
- PageLayout now changes `document.title` on render

sharvit commented on app/registries/foreman/access_permissions.rb:
```diff
@@ -756,4 +756,8 @@
   end
+
+  permission_set.security_block :react do |map|
+    map.permission :view_react, {:react => [:index]}
```
I'm confused, why is this needed?
Do we need special permission to see react pages or it should just be allowed to everyone?

glekner replied:
these changes are needed for a rake test to pass (permissions_test, seed_test)

tbrisker replied:
They shouldn't be. We don't add permissions just so tests pass when they don't make sense from a user standpoint. opened https://projects.theforeman.org/issues/27333 to remove this.

sharvit commented on webpack/assets/javascripts/react_app/common/urlHelpers.js:
```diff
@@ -13,3 +15,60 @@ export const urlBuilder = (controller, action, id = undefined) =>
+/**
+ * Get updated page param
+ */
+export const getURIpage = () => Number(getURI().query(true).page) || 1;
```
test?

sharvit commented on webpack/assets/javascripts/react_app/common/urlHelpers.js:
```diff
@@ -13,3 +15,60 @@ export const urlBuilder = (controller, action, id = undefined) =>
+/**
+ * Get updated perPage param
+ */
+export const getURIperPage = () => Number(getURI().query(true).per_page);
```
test?

sharvit commented on webpack/assets/javascripts/react_app/common/urlHelpers.js:
```diff
@@ -13,3 +15,60 @@ export const urlBuilder = (controller, action, id = undefined) =>
+/**
+ * Get updated searchQuery param
+ */
+export const getURIsearch = () => getURI().query(true).search || '';
```
test?

sharvit commented on webpack/assets/javascripts/react_app/common/urlHelpers.js:
```diff
@@ -13,3 +15,60 @@ export const urlBuilder = (controller, action, id = undefined) =>
+/**
+ * Get updated URI params
+ */
+export const getParams = () => ({
```
test?

sharvit commented on webpack/assets/javascripts/react_app/components/SearchBar/SearchBarHelpers.js:
```diff
@@ -1,8 +1,7 @@
   const data = { ...uri.query(true), search: searchQuery.trim(), page: 1 };
-  uri.query(URI.buildQuery(data, true));
-  window.Turbolinks.visit(uri.toString());
+  changeQuery(uri, data);
```
Do we have to call `getURI()` and pass it to the `changeQuery` every time we want to use it?
Would it make sense if `changeQuery` will call `getURI` instead of the consumer?

glekner replied:
Sometime consumers need to `getURI()` before calling `changeQuery` (manipulating data etc), so if we put it inside it will be called twice for no reason.

sharvit replied:
What if `changeQuery` would be `changeQuery(data, uri = getURI())`?

sharvit commented on webpack/assets/javascripts/react_app/pages/common/PageLayout/PageLayout.js:
```diff
@@ -13,54 +13,59 @@ const PageLayout = ({
-          <div className="row form-group">
-            <h1 className="col-md-8">{header}</h1>
+}) => {
+  document.title = header;
```
Can we create a small helper function for this operation (e.g. `updateDocumentTitle(header)`)?
IMO it can live here: https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/common/document.js

sharvit reviewed (changes requested):
Thanks @glekner

sharvit commented on webpack/assets/javascripts/react_app/common/__snapshots__/urlHelpers.test.js.snap:
```diff
@@ -0,0 +1,17 @@
+Object {
+  "page": 1,
+  "perPage": NaN,
+  "searchQuery": "",
```
Can we simulate real values for the tests?
Maybe by updating the `window.location`?

sharvit commented on webpack/assets/javascripts/react_app/common/urlHelpers.test.js:
```diff
@@ -21,3 +31,34 @@ describe('urlWithSearch', () => {
+    expect(getParams()).toMatchSnapshot();
+    expect(getURIpage()).toMatchSnapshot();
+    expect(getURIperPage()).toMatchSnapshot();
+    expect(getURIsearch()).toMatchSnapshot();
```
Could you give a description to your snapshots? `.toMatchSnapshot('getParams')`

sharvit reviewed (changes requested):
Thanks @glekner

glekner commented:
added tests, removed `SearchBarHelper` and `PaginationHelper`
@sharvit can you ack?

sharvit reviewed (approved):
ACK

glekner commented:
can we merge @mmoll @ohadlevy ? :)

mmoll commented:
merged, thanks @glekner!
